### ISSUE TITLE
Support async pre/post in plugins

### ISSUE THREAD

**Issue by timofei-iatsenko** [babel/babel#16860]
### 💻

- [ ] Would you like to work on this feature?

### What problem are you trying to solve?

Ecosystem moving towards native ECMAScript modules and even babel itself is [finally supporting](https://github.com/babel/babel/pull/12266) plugins/preset written in ESM. However, there still [babel-plugin-macros](https://github.com/kentcdodds/babel-plugin-macros) which is used in numerous of application which is not supporting ESM and will not without this change. 

I'm as a maintainer of [Lingui](https://github.com/lingui/js-lingui) want to modernize codebase and use native ESM modules, but support babel-macros is blocking the whole project of switching to ESM. 

The `babel-plugin-macros` is working in the way that is looking for the imports satisfying a specific pattern and then load an associated module with macro code. 

Because writing async visitor is impossible by many reasons, the logic of  `babel-plugin-macros` could be changed. 

Plugin could use regexp-based algorithm to find all modules it needs to load in the `pre` function, and then use loaded modules in the synchronous visitor. 

I've found [that discussion](https://github.com/babel/babel/discussions/11768) where it was mentioned that if there would be a good case, it might be considered to be supported. I believe unblocking module which has weekly downloads of 15 million times to use a ESM should be worth it.  

### Describe the solution you'd like

Make a `pre/post` async.

### Describe alternatives you've considered

The solutions such as 

https://github.com/addaleax/synchronous-worker
https://github.com/un-ts/synckit

Will not work when it comes to loading a ESM module, because they use structured clone algoritm, and modules could not be serialized using it.

**Comment by liuxingbaoyu** [comment]
I think we have supported them.

https://github.com/babel/babel/blob/343ee2d817b19a8cb4488c4b21f8fd6d2dcd9834/packages/babel-core/src/transformation/index.ts#L101-L102

**Comment by nicolo-ribaudo** [comment]
We have the code ready, it just need we to actually await the result of pre/post instead of throwing :)

PRs are welcome. Just note that for plugins becoming async is a breaking change, because they would stop working with `.transformSync`.

**Comment by conartist6** [comment]
I'm not working directly on fixing Babel, but in BABLR these problems disappear because even visitors can do async work without having to be async in the performance-wrecking `async function` sense.

Nicolo was quite right when he called out "waiting one tick for every visitor method would be an enormous performance penalty," so the key is not to incur that penalty except at the moments you absolutely must, and generators are the tool that lets us do it.

**Comment by conartist6** [comment]
The structure of functions gains some messy boilerplate, but you get what you want: recursive function calling where the whole stack of functions waits, but only when it has to.

```js
export const identity = (iterable) => {
   // See implementation: https://gist.github.com/conartist6/ebcb1b0b37bf24f4d75de7924443f8b2#file-index-js-L113-L129
  return  new StreamIterable(__identity(iterable));
}

const __identity = function* identity(iterable) {
  const co = iterable[Symbol.for('@@streamIterator'`];

  try {
    let step = co.advance();

    for (;;) {
      if (step instanceof Promise) {
        // yield promises instead of awaiting them
        step = yield step;
      }

      if (step.done) break;

      const instr = step.value;

      const returnValue = yield instr;

      step = co.advance(returnValue);
    }
  } finally {
    co.return();
  }

  return step.value;
};
```

**Comment by timofei-iatsenko** [comment]
@conartist6 I believe that BABLR could save that issue, but `babel-plugin-macros` is already built-in in so many tools, so we have an obligation to keep supporting it. Unblocking ESM for it would be a very big step towards modern ecosystem.

**Comment by conartist6** [comment]
Yeah, I do agree. I'm also the current maintainer of `babel-plugin-macros` as well as a heavy user of it.

What's the plan to deal with `transformSync`? Should those users just get an error message that they'll have to do an async transform to get support for ESM in `babel-plugin-macros`?

**Comment by timofei-iatsenko** [comment]
> Should those users just get an error message that they'll have to do an async transform to get support for ESM in babel-plugin-macros?

Yes. It looks the most popular case for the babel-macro-plugin is CRA which uses webpack loader already covered, because webpack-loader uses async version of babel. For everything else need error would be thrown.
